transformer_engine/pytorch/module.py

```diff
@@ -105,6 +109,10 @@ def get_workspace() -> torch.Tensor:
 def _prepare_backward(fp8: bool, fp8_meta: Dict[str, Any],  name: str = "") -> None:
     """Checks and prep for BWD."""
     if fp8:
+        global _amax_reduce_handle_bwd
+        if is_last_fp8_module() and _amax_reduce_handle_bwd is not None:
+            _amax_reduce_handle_bwd.wait()
```

Comment:
I'm not sure I fully understand this PR, but the goal is to launch an async amax reduction when doing the backward pass for the first FP8 module and synchronize before the next backward pass (i.e. the backward pass for the last FP8 module)? Could we accomplish the same behavior by resetting the request object after synchronizing?
```suggestion
        global _amax_reduce_handle_bwd
        if  _amax_reduce_handle_bwd is not None:
            _amax_reduce_handle_bwd.wait()
            _amax_reduce_handle_bwd = None
```
This would be somewhat simpler since it doesn't need the `is_last_fp8_module` infrastructure.

Reply:
I believe the last module is to keep consistent with is_first_module. I'll test this feature once.

Reply:
The goal of this PR is (1) overlap amax reudction with p2p in both pipeline warmup and flush phases, (2) overlap amax reduction with backprop (in case of the reduction after fwd microbatch) and fprop (in case of the reduction after bwd microbatch).

Doesn't your suggested change actually do sync after each Transformer backward not after the last one?

Reply:
This change is not a standalone change and `_amax_reduce_handle_bwd` would need to be set properly each time IIUC.